### ISSUE TITLE
Battery CCL ignored by MPPT control

### ISSUE THREAD

**Issue by weichenb** [victronenergy/venus#1358]
CCL defined by BMS or in DVCC should actually assure battery health.
Such definitions are properly instructed to Multiplus's, but not to MPPTs.
When CCL is set to 0 by BMS MPPTs continue charinging until having reached float voltage and are NOT properly controlled by VenusOS.
In the example below please see a case where a BYD LVL provided CCL=0, but MPPT provided up to 46A during a period of 6 minutes.
![Screenshot 2024-10-06 164459](https://github.com/user-attachments/assets/efd32636-e240-4f04-8b46-4a8e52c0d360)

I would appreciate, if MPPTs are not only controlled by CVL but also by CCL.

**Comment by thepaulcooper** [comment]
I wonder if this is why I get very frequent "**BYD Premium LV battery - Warning - High voltage**" messages where the reported voltage momentarily peaks at over 56 volts?  Weichenb, do you get the same?  My setup is a MultiPlus II GX, MPPT RS 450/100 and 12kWh of BYD LV batteries with the BYD BMS.

**Comment by weichenb** [comment]
@thepaulcooper : On my side I haven't experienced any over-voltage issues.
Setup:
- MultiPlus-II 48/5000/70-48 - 3 phase setup
- SmartSolar MPPT VE.Can 250/100 rev2
- Fronius Symo 6.0-3-M attached to AC-OUT

**Comment by asdt1803** [comment]
Is it possible, that this issue not only affects BYD batteries? I'm having similar issues with my Gobel Power batteries. 
The Pace BMS also does not feature a function to limit CCL. As the DVCC CCL does not have any affect on my RS450/200 and RS450/100 MPPT, it happens on days with highly fluctuating Solar Power, that one of the batteries goes into Cell overvoltage protection when the weakest cell of the batterypack peaks due to a suddenly high charge current. 
Unfortunately the Pace BMS transmits only the lowest battery voltage over CAN. As the voltage of the battery in protection state decreases, the Victron system continues to charge the remaining batteries. 
But: It does that without respecting the CVL-Limit configured in the DVCC setting and in the BMS settings (in my case both configured to 55,2 V)! I can measure up to 57,6 V directly at the Lynx Busbar, so that the remaining batteries enter into overvoltage protection one after another until all batteries are in protection state.
![PXL_20241004_121216034](https://github.com/user-attachments/assets/270657f5-bf2e-4bd1-aea7-fecc5760422d)
 
The DMM on the picture is connected directly to the Lynx Busbar.

Strangely the overshooting voltage is not visible in the VRM graphics even though this bevaviour sometimes occurs for more than a minute.

My System:

3 x MPII 8000
1 x RS450/200
1 x RS450/100
1 x Fronius Symo 10.0-3 in AC input
4 x Gobel Power PG-SR1-PC200 (Pace BMS)
29,92 kWp

**Comment by weichenb** [comment]
> My System:
> 
> 3 x MPII 8000 1 x RS450/200 1 x RS450/100 1 x Fronius Symo 10.0-3 in AC input 4 x Gobel Power PG-SR1-PC200 (Pace BMS) 29,92 kWp

As I can see and think to understand, I suppose that your PACE BMS is not supported by dvcc.py
```
	0xB004: _lg_quirk,
	0xB009: _pylontech_quirk(),
	0xB00A: _byd_quirk,
	0xB015: _byd_quirk,
	0xB019: _byd_quirk,
	0xB029: _pylontech_pelio_quirk,
	0xA3E4: _lynx_smart_bms_quirk,
	0xA3E5: _lynx_smart_bms_quirk,
	0xA3E6: _lynx_smart_bms_quirk,
	0xA3E7: _lynx_smart_bms_quirk,
```
In the moment only LG, BYD, Pylontech and Victron BMS seem to be supported by that script.

**Comment by asdt1803** [comment]
> Btw. what voltages did you set in Pace BMS?
> 
> In case you configured Pylontech simulation in Pace, voltage is directly taken from Pace.

The BMS uses Pylontech protocal. In the BMS-settings I left the default value of 55.2 V. 
In fact I usually set a lower voltage via CVL, especially when I'm away from home, to avoid this charge voltage overshoots.  As said, CVL usually works for me. The only trouble occurs with full or nearly full batteries in combination with highly fluctuating solar power as described above. In that case the Charge voltage exceeds the configured limits for whatever reason.

**Comment by cornim** [comment]
I have the same issue with CCL being supplied by a REC BMS.
The MPPTs actually push the current going into the battery up to a point where the battery goes into an emergency shutoff, while all CCL values (the ones from Victron under AVCC settings and the ones from the BMS) are being completely ignored. 
Just to be clear here, the grid export does not start even though the charging current for the battery is higher than the CCL value. 
Are there any work arounds for this?

**Comment by WiliceK** [comment]
I am experiencing the same issue with ESS + DC-coupled PV + managed battery (JK Inverter BMS via CAN, Victron protocol).
The system correctly receives and displays CCL from the BMS, and MPPTs show External control, but in practice the charge current is not limited by CCL. Solar chargers continue charging the battery above the CCL value for a sustained period (not short spikes), which eventually triggers BMS overcurrent protection and shutdown.
This leads to a situation where instead of exporting excess energy to the grid, the system keeps feeding the battery even when CCL is exceeded.
Setup:
- 3xMultiPlus-Ii 3000/48
- Cerbo GX with Venus OS 3.73
- 594Ah 48V Battery with JK Inverter BMS (FW 19.31, Victron CAN protocol)
- 3 MPPTs - VE.Can MPPTs (250/100, 250/80) + VE.Direct MPPT (150/70)
- Lynx Shunt for measurement
- ESS with DC feed-in enabled, no Dynamic ESS

This behaviour is reproducible and matches what others describe above.

**Comment by weichenb** [comment]
@WiliceK 
related to MPPTs VenusOS currently only evaluates CVL. I assume that CVL voltage is not or not sufficently low set in your case. Typically BMS set both parameters to reduce charging. For some specific batteries VenusOS limits CVL to 55,0V when SOC  of 100% is reached.
Please check your CVL setting that your BMS sends to lower or to stop charging.
Btw. did you activate DVCC?
More details about VenusOS control retrieve from /opt/victronenergy/dbus-systemcalc-py/delegates/dvcc.py

@mpvader Manual set max charge current in DVCC is also ignored by MPPTs. Do have a plan to change this functional weakness / bug?

<img width="612" height="542" alt="Image" src="https://github.com/user-attachments/assets/c6420b3c-ce07-45dd-976b-01811e95b38d" />

**Comment by WiliceK** [comment]
> Please check your CVL setting that your BMS sends to lower or to stop charging.
> Btw. did you activate DVCC?

CVL is not the limiting factor in my case.

The JK BMS keeps the CVL at 56.0 V until the absorption stage is complete (one hour after reaching 56.0 V). Only then does it switch to float and reduce the CVL to 54.0 V. When the SOC drops below 90%, the CVL returns to 56.0 V.

The issue occurs long before the battery reaches CVL.

Example:

Battery SOC is around 40%.
The PV power increases as the sun rises.
The MPPTs increase their output and charge the battery.
The battery reports a CCL lower than the available PV charge current.
Instead of limiting the battery charge current to the reported CCL and exporting the excess PV power to the grid, the system continues charging the battery above the CCL.
Eventually, the JK BMS trips on Charge Overcurrent Protection and disconnects charging.

In my opinion, ESS should treat the BMS-reported CCL as the maximum allowable battery charge current and redirect any excess DC-coupled PV power to grid export.

And yes, DVCC is enabled, and the system is operating with the JK BMS as a managed CAN battery.

**Comment by weichenb** [comment]
56,0V sound for me too high. LiFePo batteries have a float voltage between 53,5 und 55,0V. Some other batteries reach 100% SOC before 55V. E.g. Pylontech batteries have an absorptiom voltage of 53,6V.
Please check datasheet of your batteries

As stated MPPTs only use CVL and not CCL!
By reducing CVL charge current gets also reduced! I implemented a MPPT charge control that reduces charge current to reduce battery temperature by influencing CVL.

**Comment by WiliceK** [comment]
@mpvader I can also confirm that the manually configured DVCC Maximum Charge Current is ignored by the MPPTs when DC-coupled PV feed-in is enabled. But this is described in ESS manual and expected behavior.

However, I consider this a secondary issue.

The more critical problem is that the Charge Current Limit (CCL) received from the managed BMS also appears to be ignored under the same conditions. Unlike the manual DVCC limit, the BMS-reported CCL is a battery safety limit. Ignoring it can cause the battery charge current to exceed the maximum allowed value, eventually triggering the BMS Charge Overcurrent Protection and interrupting system operation.

Are there any plans to change this behavior so that ESS respects the BMS-reported CCL and exports excess PV power to the grid instead of continuing to charge the battery above the CCL?

**Comment by WiliceK** [comment]
> 56,0V sound for me too high. LiFePo batteries have a float voltage between 53,5 und 55,0V. Some other batteries reach 100% SOC before 55V. E.g. Pylontech batteries have an absorptiom voltage of 53,6V. Please check datasheet of your batteries
> 
> As stated MPPTs only use CVL and not CCL! By reducing CVL charge current gets also reduced! I implemented a MPPT charge control that reduces charge current to reduce battery temperature by influencing CVL.

Reducing the CVL only limits the charge current once the battery voltage approaches the CVL. My issue occurs while the battery is still far from full charge (for example, at around 40% SOC as described above), with the battery voltage still several volts below the CVL. In this operating region, charge current should be limited by the BMS-reported CCL, since the CVL has not yet become the limiting factor.

**Comment by cornim** [comment]
I can confirm everything that's been said so far. 

a) the ESS assistant ignores BMS reported CCL when DC grid feed in is active
b) this is documented behaviour 

In that sense this is a feature request and not an issue. 

Currently you can only use the ESS assistant if you're battery pack is large enough so it can take your peak PV current as charging current.
I guess this is the case because Victron is primarily geared towards offshore, RV, etc. use cases where normally the battery is much larger in comparison to the PV installation.

However the fact that this is the case was completely unexpected for me especially for a premium brand like Victron.

Bottom line from me: I had to set my inverters to external control and build the entire charge discharge control logic myself from scratch. Since this took a significant amount of time I will not ever buy Victron again or recommend it to anybody unless this is fixed.

**Comment by WiliceK** [comment]
> a) the ESS assistant ignores BMS reported CCL when DC grid feed in is active
> b) this is documented behaviour

I don't think this behavior is actually documented.

The documentation explicitly states that the manual DVCC Maximum Charge Current is ignored when DC-coupled PV feed-in is enabled. However, I could not find any documentation stating that the BMS-reported CCL is also ignored under the same conditions.

Moreover, in this discussion (https://communityarchive.victronenergy.com/questions/187320/dvcc-feedin.html), Nick de Boer (Victron) stated:

"The manual override ... is ignored. Expected and documented. The actual CCL sent by the BMS ... should be honoured."

This suggests that the documented exception only applies to the manual DVCC current limit, while the CCL provided by a managed BMS is still expected to be respected.

If that understanding is correct, then the behavior described in this issue appears to be inconsistent with both the documentation and the clarification above.

**Comment by WiliceK** [comment]
To summarize, I believe the expected ESS behavior with DC-coupled PV feed-in and a managed CAN BMS should be:

1. MPPTs should continue operating at maximum available PV power.
2. The battery charge current should never exceed the CCL reported by the BMS.
3. Any PV power exceeding the BMS-reported CCL should be exported to the grid (if grid export is enabled), rather than charging the battery above its allowed current.

This is the only behavior that is consistent with the purpose of a managed battery. The BMS-reported CCL is not an optimization target but a battery safety limit. If ESS ignores it, batteries that regulate charging primarily via CCL can repeatedly trip on Charge Overcurrent Protection, even though sufficient grid export capacity is available.

**Comment by weichenb** [comment]
> Reducing the CVL only limits the charge current once the battery voltage approaches the CVL.

With a reduced voltage I 
- even prevent that that battery is fully charged by MPPTs (intend to increase battery lifetime - see screenshot) and I
- can limit the current to avoid that battery unnecessarily gets hot while being charged with high currents

If you instruct BMS to provide lower CVS for such cases when lower CCLs are expected, then you will also get expected behaviour.
However, you should consider to select not an too low voltage. In that cases battery gets discharged and power goes into grid.

<img width="456" height="547" alt="Image" src="https://github.com/user-attachments/assets/2018872c-8a6a-425a-ac64-1f0ff7ffe7a5" />

**Comment by drurew** [comment]
Proposed fix at victronenergy/dbus-systemcalc-py#23. Single-line change in `set_maxchargecurrent`: replace `maximize_charge_current()` with `_set_charge_current(solarchargers, max_charge_current)` when feedback is allowed. The Multi-RS path already does this correctly.
